### PR TITLE
SymmetryGroup Pass-through for Volume Transformations

### LINKED ISSUE #1081: Volume transformation removes `symmetry_group`
Transforming a `Volume` via `flip`, `downsample`, `rotate`, `transpose`, etc. ... removes `symmetry_group` information.

### PR DESCRIPTION
**Pull request by j-c-c**
Checks whether operations on symmetric volumes should retain symmetry group. If not, the symmetry group is downgraded to C1 and warning is emitted.

Will close #1081.

### CODE REVIEW

File: src/aspire/volume/volume.py [hunk 1]
```diff
@@ -176,6 +179,51 @@ def _set_symmetry_group(self, value):
...
+        warnings.warn(msg, UserWarning, stacklevel=4)
+
+    def _result_symmetry(self, other=None):
+        """
+        Check if `other` will alter symmetry of `self` and return resulting symmetry_group.
+
+        :return: SymmetryGroup instance.
+        """
+        # No need to check for C1.
+        if str(self.symmetry_group) == "C1":
```
Comment (garrettwrong) [inline src/aspire/volume/volume.py]: Why not check if it is IdentitySymmetryGroup directly?

File: src/aspire/volume/volume.py [hunk 2]
```diff
@@ -176,6 +179,51 @@ def _set_symmetry_group(self, value):
...
+        # Handle case of transformations of `self`.
+        if other is None:
+            self._symmetry_group_warning()
+            return IdentitySymmetryGroup(dtype=self.dtype)
+
+        # Handle binary operations of `self` and `other`.
+        result_symmetry = self.symmetry_group
+
+        # Warn and set to IdentitySymmetryGroup if `other` is symmetrically incompatible.
+        if isinstance(other, Volume) and str(self.symmetry_group) != str(
```
Comment (garrettwrong) [inline src/aspire/volume/volume.py]: We can give `SymmetryGroup` classes a notion of equality.

File: src/aspire/volume/volume.py [hunk 3]
```diff
@@ -176,6 +179,51 @@ def _set_symmetry_group(self, value):
...
+        # No need to check for C1.
+        if str(self.symmetry_group) == "C1":
+            return self.symmetry_group
+
+        # Handle case of transformations of `self`.
+        if other is None:
+            self._symmetry_group_warning()
+            return IdentitySymmetryGroup(dtype=self.dtype)
+
+        # Handle binary operations of `self` and `other`.
```
Comment (garrettwrong) [inline src/aspire/volume/volume.py]: This comment reads like it was for a different bit of code; or it is just confusing to me.

File: src/aspire/volume/volume.py (hunk 2 above)
Comment (garrettwrong) [inline src/aspire/volume/volume.py]: try seeing what this look like as the following (but with better names than cond1)

```
cond1 = ...
cond2 = ...
cond3 = ...

result_symmetry = self.result_symmetry
if any(cond1,cond2,cond...):
    self._symmetry_group_warning()
    result_symmetry = IdentitySymmetryGroup(dtype=self.dtype)
```

**Review by garrettwrong — CHANGES_REQUESTED** [review]
Thanks for putting this together.  Couple things to try that might tidy this up. Not sure if end result will be nicer yet, but probably worth trying at this stage.

File: src/aspire/volume/volume.py (hunk 1 above)
Comment (j-c-c) [reply src/aspire/volume/volume.py]: oof, yeah. done.

File: src/aspire/volume/volume.py (hunk 3 above)
Comment (j-c-c) [reply src/aspire/volume/volume.py]: Removed.

File: src/aspire/volume/volume.py (hunk 2 above)
Comment (j-c-c) [reply src/aspire/volume/volume.py]: Added override of `__eq__` for `SymmetryGroup`s. As is, I am allowing symmetry groups of differing `dtype` to be equivalent. Can change this if you think it will be problematic.

Comment (j-c-c) [reply src/aspire/volume/volume.py]: Much cleaner thanks!

File: src/aspire/volume/volume.py [hunk 4]
```diff
@@ -176,6 +179,44 @@ def _set_symmetry_group(self, value):
...
+
+        # Warn and set to IdentitySymmetryGroup if `other` is symmetrically incompatible.
+        result_symmetry = self.symmetry_group
+
+        # Conditions of incompatibility.
+        self_transformation = other is None
+        incompat_syms = (
+            isinstance(other, Volume) and self.symmetry_group != other.symmetry_group
+        )
+        arbitrary_array = not isinstance(other, Volume) and hasattr(other, "__len__")
```
Comment (garrettwrong) [inline src/aspire/volume/volume.py]: what is the purpose of checking for "__len__"?  Maybe there is a better way to do this.

Also consider x = np.array([123]) which is a scalar, but has a length...

File: tests/test_volume.py [hunk 5]
```diff
@@ -726,6 +734,77 @@ def test_symmetry_group_set_get(sym_group, sym_string):
...
+
+def test_transformation_symmetry_warnings(symmetric_vols):
+    """
+    A warning should be emitted (once) for transpose, flip, rotate, add, sub, mult, div.
+    """
+    vol_c3, _ = symmetric_vols
+    sym_group = str(vol_c3.symmetry_group)
+    assert sym_group == "C3"
+
+    # Check we get warning on first transformation.
```
Comment (garrettwrong) [inline tests/test_volume.py]: Not sure I follow this comment.  Is the code below testing for 3 warnings?

File: tests/test_volume.py [hunk 6]
```diff
@@ -726,6 +734,77 @@ def test_symmetry_group_set_get(sym_group, sym_string):
...
+        str(vol_c3.stack_reshape((1, 1)).symmetry_group) == sym_group
+    )  # stack_reshape
+    assert (
+        str(vol_c3.downsample(vol_c3.resolution // 2).symmetry_group) == sym_group
+    )  # downsample
+
+
+def test_transformation_symmetry_warnings(symmetric_vols):
+    """
+    A warning should be emitted (once) for transpose, flip, rotate, add, sub, mult, div.
```
Comment (garrettwrong) [inline tests/test_volume.py]: The algebraic ops are in a diff function, snip this part of docstring into the other function.

File: tests/test_volume.py [hunk 7]
```diff
@@ -726,6 +734,77 @@ def test_symmetry_group_set_get(sym_group, sym_string):
...
+    assert str(vol_r.symmetry_group) == "C1"
+
+    # Check original volume has retained C3 symmetry.
+    assert str(vol_c3.symmetry_group) == "C3"
+
+
+def test_aglebraic_ops_symmetry_warnings(symmetric_vols):
+    vol_c3, vol_c4 = symmetric_vols
+
+    # Compatible symmetry should retain symmetry_group.
```
Comment (garrettwrong) [inline tests/test_volume.py]: Might want to programtically check that warnings are not emitted here.

File: tests/test_volume.py [hunk 8]
```diff
@@ -726,6 +734,77 @@ def test_symmetry_group_set_get(sym_group, sym_string):
...
+        vol_array_div = vol_c3 / (vol_c4.asnumpy() + 1)
+
+    assert str(vols_sum.symmetry_group) == "C1"
+    assert str(vol_array_diff.symmetry_group) == "C1"
+    assert str(vols_mult.symmetry_group) == "C1"
+    assert str(vol_array_div.symmetry_group) == "C1"
+
+    # Should have 4 warnings on record.
+    assert len(record) == 4
+
```
Comment (garrettwrong) [inline tests/test_volume.py:812]: Where are you checking that the warning only occurs at most once?  It looks like this only checks that each warning occurs once per one call.  Maybe this is what you meant, but I thought we talked about something to the effect of:

```
for i in lots:
   # stuff
   v = v + npv
```

only triggering only once per that line.  If that is the behavior, I'm not sure this test covers.

**Review by garrettwrong — CHANGES_REQUESTED** [review]
Cool thanks for updating, looked over the tests this time.  Pretty close! Let me know if you have any questions. Thanks

File: src/aspire/volume/volume.py (hunk 4 above)
Comment (j-c-c) [reply src/aspire/volume/volume.py]: What do you think about `np.array(other).size > 1`?

Comment (garrettwrong) [reply src/aspire/volume/volume.py]: Python has a way to get attr if it exists with a default when it doesn't.  Probably can just use that and get the situations that matter here.

https://docs.python.org/3/library/functions.html#getattr

Comment (j-c-c) [reply src/aspire/volume/volume.py]: It appears using `getattr(np.array([123]), '__len__')` returns a method wrapper object instead of the actual length. I could follow it with a call, `getattr(np.array([123]), '__len__')()`, to get the length, but that wouldn't work for actual scalars.

Comment (j-c-c) [reply src/aspire/volume/volume.py]: I'll try to come up with something else.

Comment (garrettwrong) [reply src/aspire/volume/volume.py]: you dont need to use __len__

Comment (j-c-c) [reply src/aspire/volume/volume.py]: Ah yeah, ok. Using `getattr(other, size, 1) > 1` does the trick. Thanks.

File: tests/test_volume.py (hunk 6 above)
Comment (j-c-c) [reply tests/test_volume.py]: oops. fixed. thanks!

File: tests/test_volume.py (hunk 5 above)
Comment (j-c-c) [reply tests/test_volume.py]: Changed this comment. Yes, we are expecting 3 warnings here.

File: tests/test_volume.py (hunk 7 above)
Comment (j-c-c) [reply tests/test_volume.py]: Added a check for this.

File: tests/test_volume.py (hunk 8 above)
Comment (j-c-c) [reply tests/test_volume.py:812]: This is the default behavior for warnings and I have confirmed that we only get one warning per line, however I have not found a nice way to test this. `pytest.warns()` seems to pick up all the warnings, including those that would be suppressed.

Comment (j-c-c) [reply tests/test_volume.py:812]: `warnings.catch_warnings()` did the trick here. Check added.

**Review by garrettwrong — DISMISSED** [review]
Great, I think you covered all the remarks from the first pass, thanks!

File: src/aspire/volume/volume.py [hunk 9]
```diff
@@ -176,6 +179,46 @@ def _set_symmetry_group(self, value):
...
+        """
+        # No need to check for C1.
+        if isinstance(self.symmetry_group, IdentitySymmetryGroup):
+            return self.symmetry_group
+
+        # Warn and set to IdentitySymmetryGroup if `other` is symmetrically incompatible.
+        result_symmetry = self.symmetry_group
+
+        # Conditions of incompatibility.
+        self_transformation = other is None
```
Comment (janden) [inline src/aspire/volume/volume.py]: Why should self transformations always trigger reset to C1?

File: src/aspire/volume/volume.py [hunk 10]
```diff
@@ -350,16 +400,21 @@ def from_vec(cls, vec):
 
         return cls(data)
 
-    def transpose(self):
+    def transpose(self, symmetry=None):
```
Comment (janden) [inline src/aspire/volume/volume.py]: What's the use case for supplying a `symmetry` argument here? Also, it should probably be documented.

File: src/aspire/volume/volume_synthesis.py [hunk 11]
```diff
@@ -247,6 +248,9 @@ def _check_order(self):
                 f"An {self.__class__.__name__} must have order=1. Provided order was {self.order}"
             )
 
+    def _set_symmetry_group(self):
```
Comment (janden) [inline src/aspire/volume/volume_synthesis.py:251]: Why is this added here?

**Review by janden — CHANGES_REQUESTED** [review]
Looks great. Just a few things.

File: src/aspire/volume/volume.py (hunk 9 above)
Comment (j-c-c) [reply src/aspire/volume/volume.py]: This covers the cases of `rotate`, `transpose`, and `flip`; the "self-transformations" that do not maintain symmetry axes. The only self transformation that retains the symmetry group is `downsample`, which doesn't call `_result_symmetry` it just passes along `self.symmetry_group`.

File: src/aspire/volume/volume.py (hunk 10 above)
Comment (j-c-c) [reply src/aspire/volume/volume.py]: The purpose of this is to ensure that the warning emitted by `self._result_symmetry()` happens at the same stack level whether a user calls `vol.T` or `vol.transpose()`. This way the traceback points to location in the code where `T` or `transpose` was called.

Seeing as I don't really want a user setting this parameter I should probably just use a private flag and remove this argument altogether. I'll fix this up.

File: src/aspire/volume/volume_synthesis.py (hunk 11 above)
Comment (j-c-c) [reply src/aspire/volume/volume_synthesis.py:251]: Prior to adding this, `AsymmetricVolume`'s were instantiated with `CnSymmetryGroup` of `order = 1`. All other asymmetric volumes in the code were using `IdentitySymmetryGroup`, for example wrapping a numpy array in `Volume` or downloading `emdb_2660` using the data downloader. This was just to enforce that all asymmetric volumes were using the `IdentitySymmetryGroup`. I realized this was the case when I tried to use the symmetry group equality operator `__eq__` (added in this PR) on two asymmetric volumes instantiated in different ways.

File: src/aspire/volume/volume.py (hunk 10 above)
Comment (j-c-c) [reply src/aspire/volume/volume.py]: Handling this with a private flag now. Let me know what you think.

File: src/aspire/volume/volume.py (hunk 9 above)
Comment (janden) [reply src/aspire/volume/volume.py]: Wasn't this supposed to be renamed?

Comment (j-c-c) [reply src/aspire/volume/volume.py]: Yep, sorry about that. I think I tagged you yesterday before the meeting. I was waiting for CI on another branch so I could update this. 

Changed the name to `axes_altering_transformation`

**Review by janden — APPROVED** [review]
Sorry I missed that this was ready. Here you go!

**Comment by j-c-c** [comment]
> Sorry I missed that this was ready. Here you go!

Thanks!
